File: FAANG/Binary Trees/MaxDepthBinaryTree/MaxDepthBinaryTree.java
package FAANG.Binary Trees.MaxDepthBinaryTree;

public class MaxDepthBinaryTree {
  
  class BNode {

    public int value;
    public BNode left;
    public BNode right ;

    public BNode(int value,BNode left,BNode right){
        this.value = value;
        this.left = left;
        this.right = right;
    }

    public BNode(int value){
        this.value = value;
        this.left = null;
        this.right = null;
    }
  }
  public static int maxDepth(BNode node, int currentDepth){
    if (node == null ) return  currentDepth;
    currentDepth++;
    return  Math.max(maxDepth(node.left, currentDepth),
            maxDepth(node.right,currentDepth));
  }

  public BNode buildTree(int[] arr, int rootIndex) {
      BNode root = null ;
      // Base case for recursion
      if(rootIndex<arr.length){
          root = new BNode(arr[rootIndex]);
          root.left = buildTree(arr,(2*rootIndex)+1);
          root.right = buildTree(arr,(2*rootIndex)+2);
      }
      return root;
  }
  public  void printTree(BNode root){
      if(root!=null){
          printTree(root.left);
          System.out.print(root.value+" ");
          printTree(root.right);
      }
  }

  public static void main(String[] args) {
      int[] array = new int[]{1,2,3,4,5,6};
      DepthOfBinaryTree depthOfBinaryTree = new DepthOfBinaryTree();
      BNode btree = depthOfBinaryTree.buildTree(array,0);
    depthOfBinaryTree.printTree(btree);
    System.out.println("\n Max depth of the tree is "+ depthOfBinaryTree.maxDepth(btree,0));

  }
}

File: FAANG/Binary Trees/MaxDepthBinaryTree/MaxDepthBinaryTree.js
// Given the root of a binary tree, return its maximum depth.

// A binary tree's maximum depth is the number of nodes along the longest path from the root node down to the farthest leaf node.

// Example 1:
// Input: root = [3,9,20,null,null,15,7]
// Output: 3

// Example 2:
// Input: root = [1,null,2]
// Output: 2



const maxDepth = function(root) {
  if(!root) return null;
  
  return Math.max(maxDepth(root.right), maxDepth(root.left)) + 1;
};//T:O(n), S:O(n)


/*
NOTE: The beginning portion builds our test case binary tree. Scroll down to the section titled Our Solution for the code solution!
 */

// ------- Code to generate our binary tree -------
class TreeNode {
  constructor(value) {
    this.value = value;
    this.left = null;
    this.right = null;
  }

  insert(values) {
    const queue = [this];
    let i = 0;
    while (queue.length > 0) {
      let current = queue.shift();
      for (let side of ["left", "right"]) {
        if (!current[side]) {
          if (values[i] !== null) {
            current[side] = new TreeNode(values[i]);
          }
          i++;
          if (i >= values.length) return this;
        }
        if (current[side]) queue.push(current[side]);
      }
    }
    return this;
  }
}

const root = new TreeNode();
root.insert([1,1,1,1,null,null,null,1,null,null,null,1,null,null]);

// ------- Code to generate our binary tree -------

// ------- Our Solution -------
var maxDepth2 = function(node, currentDepth) {
    if (!node) {
      return currentDepth;
    }
    
    currentDepth++;
    
    return Math.max(maxDepth(node.right, currentDepth), maxDepth(node.left, currentDepth));
};

console.log(maxDepth(root, 0));